pulp_container/app/registry_api.py

```diff
@@ -475,6 +480,138 @@ def get_queryset(self, *args, **kwargs):
+    permission_classes = []
+
+    def get(self, request):
+        req_repositories = None
+        req_tags = None
+        req_oss = None
+        req_architectures = None
+        req_annotations = False
+        req_label_exists = set()
+        req_label_values = {}
+        for key, values in request.query_params.lists():
+            if key == "repository":
+                req_repositories = values
+            elif key == "tag":
+                req_tags = values
+            elif key == "os":
+                req_oss = values
+            elif key == "architecture":
+                req_architectures = values
+            elif key.startswith("annotation:"):
+                if key.endswith(":exists") and any(v != "1" for v in values):
+                    raise ParseError()
+                req_annotations = True
+            elif key.startswith("label:"):
+                if key.endswith(":exists"):
+                    if any(v != "1" for v in values):
+                        raise ParseError()
+                    label = key[len("label:") : len(key) - len(":exists")]
+                    req_label_exists.add(label)
+                else:
+                    label = key[len("label:") :]
+                    req_label_values[label] = values
+            else:
+                raise ParseError()
+
+        if "org.flatpak.ref" not in req_label_exists:
+            raise ParseError()
+
+        results = []
+
+        # This implementation never announces any annotations, so any requested annotations cause
+        # empty results:
+        if not req_annotations:
+            distributions = models.ContainerDistribution.objects.filter(private=False).only(
+                "base_path"
+            )
+
+            if req_repositories:
+                distributions = distributions.filter(base_path__in=req_repositories)
+
+            for distribution in distributions:
+                images = []
+                if distribution.repository:
+                    repository_version = distribution.repository.latest_version()
+                elif distribution.repository_version:
+                    repository_version = distribution.repository_version
+                tags = models.Tag.objects.filter(pk__in=repository_version.content)
+                manifests = models.Manifest.objects.filter(
+                    pk__in=repository_version.content,
+                    media_type__in=(models.MEDIA_TYPE.MANIFEST_V2, models.MEDIA_TYPE.MANIFEST_OCI),
+                )
+                for manifest in manifests:
+                    with storage.open(manifest.config_blob._artifacts.get().file.name) as file:
+                        raw_data = file.read()
+                    config_data = json.loads(raw_data)
+                    labels = config_data.get("config", {}).get("Labels")
+                    if not labels:
+                        continue
+                    if any(label not in labels.keys() for label in req_label_exists):
```

Comment:
it seems like you require here all specified labels to be present compared to os,arch,tags were you satisfy at least 1 match, is my read correct? and why such difference in logic only for labels?

Reply:
Oops, thanks for catching that.  Done with <https://github.com/pulp/pulp_container/commit/f5594407655fd97635edd1fdf9acd873853137fc> "Fix logic also for repeated label query parameters".

Reply:
Oh my, now I was confused.  According to <https://github.com/stbergmann/flagstate/commit/a0e3b3086db37044cd9052b417ed6fea80569e5f> "Clarify how query parameters are matched", the original code was exactly matching the clarified specification.  For `?label:l1:exists=1&label:l2:exitst=1&tag=t1&tag=t2` to match: (`label:t1:exists=1` must match "the image has a label `l1`") AND (`label:t2:exitsts=1` must match "the image has a lable `l2`") AND (`tag=t1&tag=t2` must match "the image has at least one of the tags `t1`, `t2`").

Reply:
Fixed again with <https://github.com/pulp/pulp_container/pull/1301/commits/a51d0e452fa0a0592d056f5709e43fb2a194fa82> "Revert 'Fix logic also for repeated label query parameters'".